crates/accelerate/src/basis/basis_translator/compose_transforms.rs

```diff
@@ -0,0 +1,206 @@
+            Ok(Self(data_extract))
+        } else {
+            Err(PyTypeError::new_err(
+                "Provided object was not an instance of QuantumCircuit",
+            ))
+        }
+    }
+}
+
+impl IntoPy<PyObject> for CircuitRep {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        QUANTUM_CIRCUIT
+            .get_bound(py)
+            .call_method1("_from_circuit_data", (self.0,))
+            .unwrap()
+            .unbind()
+    }
+}
+
+impl ToPyObject for CircuitRep {
+    fn to_object(&self, py: Python<'_>) -> PyObject {
+        self.clone().into_py(py)
+    }
+}
+
+#[pyfunction(name = "compose_transforms")]
+pub(super) fn py_compose_transforms(
+    py: Python,
+    basis_transforms: BasisTransforms,
+    source_basis: HashSet<(String, u32)>,
+    source_dag: &DAGCircuit,
+) -> PyResult<MappedTransforms> {
+    compose_transforms(py, &basis_transforms, &source_basis, source_dag)
+}
+
+pub(super) fn compose_transforms<'a>(
+    py: Python,
+    basis_transforms: &'a BasisTransforms,
+    source_basis: &'a HashSet<(String, u32)>,
+    source_dag: &'a DAGCircuit,
+) -> PyResult<MappedTransforms> {
+    let example_gates = *get_example_gates(py, source_dag, None)?;
+    let mut mapped_instructions: MappedTransforms = HashMap::new();
+
+    for (gate_name, gate_num_qubits) in source_basis.iter().cloned() {
+        // Need to grab a gate instance to find num_qubits and num_params.
+        // Can be removed following https://github.com/Qiskit/qiskit-terra/pull/3947 .
+        let example_gate = &example_gates[&(gate_name.clone(), gate_num_qubits)];
+        let num_params = example_gate.params_view().len();
+
+        let placeholder_params: SmallVec<[Param; 3]> = PARAMETER_VECTOR
+            .get_bound(py)
+            .call1((&gate_name, num_params))?
+            .extract()?;
+
+        let mut dag = DAGCircuit::new(py)?;
+        // Create the mock gate and add to the circuit, use Python for this.
+        let qubits = QUANTUM_REGISTER.get_bound(py).call1((gate_num_qubits,))?;
+        dag.add_qreg(py, &qubits)?;
+
+        let gate = GATE.get_bound(py).call1((
+            &gate_name,
+            gate_num_qubits,
+            placeholder_params
+                .iter()
+                .map(|x| x.clone_ref(py))
+                .collect::<SmallVec<[Param; 3]>>(),
+        ))?;
+        let gate_obj: OperationFromPython = gate.extract()?;
+        let qubits: Vec<Qubit> = (0..dag.num_qubits() as u32).map(Qubit).collect();
+        dag.apply_operation_back(
+            py,
+            gate_obj.operation,
+            &qubits,
+            &[],
+            if gate_obj.params.is_empty() {
+                None
+            } else {
+                Some(gate_obj.params)
+            },
+            gate_obj.extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            Some(gate.into()),
+        )?;
+        mapped_instructions.insert(
+            (gate_name.clone(), gate_num_qubits),
```

Comment:
Do we need to clone `gate_name` here, or can we just consume it? I don't see anymore uses after this :)